Feature/start character command

asmodee-bellanie opened the pull request:
Starts the character command.

Commit:
feat(Message Messenger): Strips the answer away.

Commit:
feat(Character): WIP, Struggles to display a message.

If character is found.
Or Not.

Commit:
feat(Character): Makes the findOne function work.

Commit:
feat(CharacterCommand): WIP, Tries to get the character name.

Commit:
feat(Mongoose): Actually calls the DB.

Commit:
Merge remote-tracking branch 'origin/develop' into feature/start-character-command

Commit:
Merge remote-tracking branch 'origin/feature/ready' into feature/start-character-command

Commit:
feat(Services): Moves the services to the right spot.

Commit:
fix(File paths): Corrects paths.

Commit:
feat(Character): WIP, tries to fix my mess.

Commit:
Merge branch 'develop' into feature/start-character-command

Commit:
feat(Prefixes): CHanges the JSON to TS.

Commit:
feat(Character creation): Makes it work.

Commit:
Merge remote-tracking branch 'origin/develop' into feature/start-character-command

C0ZEN commented on src/data/models/character.ts:
```diff
@@ -0,0 +1,40 @@
+import mongoose from 'mongoose'
```
I think that you should be careful with models name.
You have for example the User model for mongo but also a User model for TS.
You have to distinguish them when naming the files at least.
For example you could name this file character-shcema.ts and the TS model just character.ts

asmodee-bellanie replied:
Will do!

C0ZEN commented on src/features/commands/character/character.ts:
```diff
@@ -0,0 +1,25 @@
+import { CharacterService } from '../../../services/character/character.service'
+import { MessageManagerService } from '../../../services/message-manager/message-manager.service'
+
+export class CharacterCommand {
```
The class name is not matching the file name.

asmodee-bellanie replied:
Will do!

C0ZEN commented on src/features/commands/character/character.ts:
```diff
@@ -0,0 +1,25 @@
+    return CharacterCommand._instance
+  }
+
+  public displayCharacterMessage(message: Message) {
```
Since you are already in the character command class you could just sum up the method name to message.
public message(...)

asmodee-bellanie replied:
Will do!

Commit:
feat(Files): Renames files more properly.

Commit:
feat(Character command): Simplifies a method name.

C0ZEN commented on src/features/prefixes.ts:
```diff
@@ -0,0 +1,6 @@
+export enum Prefixes {
+  "!hellforge",
```
This is not a good enum.
Because you did not specify the values they will be set from 0 to X accordingly.
And once compiled an enum with numbers is able to do a reverse mapping.
Meaning that the length of your enum will double.
I do not think that this is your wish and also this is risky.
So I suggest you to do:
HELLFORGE = '!hellforge', ...

asmodee-bellanie replied:
Is an interface something I can do? :o

C0ZEN replied:
Nope.
If you want to create a list of something you can:
- Create an enum
- Create an array

C0ZEN commented on src/services/character/character.service.ts:
```diff
@@ -0,0 +1,30 @@
+    return CharacterService._instance
+  }
+
+  public async getCharacter(id: string) {
```
get is enough.
**Edit:**
I think that get is reserved since this is a keyword.
I suggest getEntity

asmodee-bellanie replied:
Get is a "protected" name, right?

C0ZEN replied:
@asmodee-bellanie check the edit

asmodee-bellanie replied:
Will do!

C0ZEN commented on src/services/character/character.service.ts:
```diff
@@ -0,0 +1,30 @@
+    })
+  }
+
+  public setCharacter(id: string, characterToInsert: CharacterType) {
```
setEntity?

asmodee-bellanie replied:
Will do!

C0ZEN commented on src/services/mongoose-connect/mongoose-connect.service.ts:
```diff
@@ -0,0 +1,27 @@
+import mongoose from "mongoose"
+import { login, password } from "../../../credentials.json"
+
+export class MongooseConnect {
```
Missing suffix service

asmodee-bellanie replied:
Will do!

C0ZEN commented on src/types/character/character-type.ts:
```diff
@@ -0,0 +1,17 @@
+  inventorySpace: number,
+  inventoryMax: number,
+  inventory: Item[],
+  ailments: Effect[]
```
ailments: what a strange word :o
why not effects?

asmodee-bellanie replied:
Will do!

asmodee-bellanie replied:
https://www.wordreference.com/enfr/ailment
Ailment is the English name for what happens to your character during the game that is not their common state. Like poisonned, paralized, stunned, etc.

It's just an effect, but it's applied against a character... whereas effects on an item are applied on the user, not the item itself.

Is my explaination convincing? Or did I get too far? xD

C0ZEN replied:
@asmodee-bellanie this is your code.
I did not knew the word but anyway you have the final word.
You do what you want and wording is clearly your choice.

asmodee-bellanie replied:
I renamed it anyway. For the beauty of consistency! XD

C0ZEN commented on src/types/character/stat.ts:
```diff
@@ -0,0 +1,5 @@
+export type Stat = {
+  name: 'Health Points' | 'Ætheris Points' | 'Strength' | 'Dexterity' | 'Defense' | 'Reflexes' | 'Senses' | 'Charisma' | 'Mind' | 'Spirit' | 'Willpower' | 'Luck' ,
```
Better to create an enum like StatNamesEnum or StatNameEnum

asmodee-bellanie replied:
Will do!

C0ZEN reviewed (changes requested):
First review

Commit:
fix(Mongoose Connect): Adds service.

Commit:
feat(Prefixes): Changes the prefixes into enum.

Commit:
feat(Character type): Changes how stats are handled.

Commit:
fix(Message manager): Changes the oterator.

Commit:
fix(Character): Changes the method names.

C0ZEN commented on src/enums/stat-names.enum.ts:
```diff
@@ -0,0 +1,14 @@
+export enum StatNamesEnum {
+  healthPoints = 'Health Points',
+  AethrisPoints = 'Ætheris Points',
```
This is not the same case as the other members.
Also (like the constants by the way), enums are usually written in SNAKE_SCREAM_CASE

asmodee-bellanie replied:
Will do!

C0ZEN commented on src/features/prefixes.ts:
```diff
@@ -0,0 +1,6 @@
+export enum PrefixesEnum {
+  long = "!hellforge",
```
Change the case?

asmodee-bellanie replied:
Will do!

C0ZEN commented on src/features/prefixes.ts:
```diff
@@ -0,0 +1,6 @@
+export enum PrefixesEnum {
```
prefixes but what for?
there could be a prefixes for almost anything so there is a lack of context here.
A prefix for what?

asmodee-bellanie replied:
Will do!

C0ZEN commented on src/services/character/character.service.ts:
```diff
@@ -0,0 +1,30 @@
+  }
+
+  public async getEntity(id: string) {
+    return await characters.findOne({ ownerId: id }, (err: Error, characterFound) => {        
```
Could you try to remove the return and test the functionality?
I am not sure that the return is required but if you could test it for me it will be nice!

asmodee-bellanie replied:
I'll try to do!

asmodee-bellanie replied:
Typescript says: an async/await function must return a value of type "Promise\<T>" if I remove the return.

C0ZEN replied:
ok this make sense

C0ZEN commented on src/services/character/character.service.ts:
```diff
@@ -0,0 +1,30 @@
+    })
+  }
+
+  public setEntity(id: string, characterToInsert: CharacterType) {
```
You should start adding return types to ALL your methods.
You will force your code to follow the safe guard type about what it should return and it will help a lot to spot the errors during development.

asmodee-bellanie replied:
Will do... ><

C0ZEN commented on src/types/character/character-type.ts:
```diff
@@ -0,0 +1,17 @@
+import { Item } from '../shop/item-type'
+import { Effect } from '../skills/effect-type'
+
+export type CharacterType = {
```
Like interfaces, adding type as suffix is bad.
Because interfaces and types are almost the same stuff except a type should be an alias or a merge of multiple interface and primitive types.
So the right thing to do here is:
- change type to interface
- remove type as suffix
- define if you want to use a convention when naming your types and interfaces and this convention should be the same one for both (ex: add model as suffix, add I as prefix, do nothing, etc...) I use I as prefix on my hand.

asmodee-bellanie replied:
Will do!

C0ZEN reviewed (changes requested):
Added more comments.

Commit:
fix(CHaracter comand): Renames the get method use.

Commit:
fix(Bot prefixes): Renames the enum and values.

Commit:
fix(Gitkeep): Removes useless file.

Commit:
fix(Stat names enum): Changes the case.

Commit:
fix(Return types): Adds proper return types.

Commit:
fix(Types): Changes types into interfaces.

C0ZEN commented on src/enums/stat-names.enum.ts:
```diff
@@ -0,0 +1,14 @@
+export enum StatNamesEnum {
+  HEALTHPOINTS = 'Health Points',
+  AETHERISPOINTS = 'Ætheris Points',
```
SNAKE_SCREAM_CASE => a "_" is missing instead of nothing to mark the space

asmodee-bellanie replied:
I will make this reptile suffer.

C0ZEN commented on src/features/commands/character/character-command.ts:
```diff
@@ -0,0 +1,25 @@
+import { CharacterService } from '../../../services/character/character.service'
+import { MessageManagerService } from '../../../services/message-manager/message-manager.service'
+
+export class CharacterCommand {
```
Service suffix is missing, right?

asmodee-bellanie replied:
Right. I need to add it.

C0ZEN commented on src/services/character/character.service.ts:
```diff
@@ -0,0 +1,30 @@
+    return CharacterService._instance
+  }
+
+  public async getEntity(id: string):Promise<Document | null> {
```
Are you sure about null?
I am not.
Test to remove null and see if there is a typing error.
Since you wrap the character inside a if, nothing is returning null.
Also throwing an error in the code should be typed as well with never.
So the minimum here should be Document | never.
If there is an error without null, reach me to talk about it.

asmodee-bellanie replied:
I will try to change the type of that function.

asmodee-bellanie replied:
Yes, there is a typing error that a if doesn't remove, sadly enough... And typing with never doesn't work.

C0ZEN replied:
Well ok let just skip this.

C0ZEN commented on src/services/character/character.service.ts:
```diff
@@ -0,0 +1,30 @@
+    characters.update({ ownerId: id }, characterToInsert, (err: Error, characterUpdated: any): Query<ICharacter> => {
+      if (err) throw err
+
+      return characterUpdated
```
How this return is useful at all?

asmodee-bellanie replied:
But then I would have an unused declared and obligatory variable... How could I do something about it? I really don't need it there... x)

asmodee-bellanie replied:
Ah shoot! We're not even obliged to set the equivalent of 'characterUpdated' on an update function ! °°

C0ZEN commented on src/types/character/character-type.ts:
```diff
@@ -0,0 +1,17 @@
+import { IStat } from './stat-type'
```
The file name is suffixed by -type.
I do not think that this is good since the exported member should match the file resulting to add back the Type suffix on the interface name which we already defined that this is not good as well.
On the other hand if you like identify your exported members you could simply put this one into a folder called either interfaces or called models or called types.

asmodee-bellanie replied:
If you look at the whole file path, there already is a 'types/' folder as a parent.

I will remove the -type from the file names.

C0ZEN commented on src/types/character/stat-type.ts:
```diff
@@ -0,0 +1,7 @@
+import { StatNamesEnum } from '../../enums/stat-names.enum';
```
Same as above about file name.

C0ZEN commented on src/types/shop/item-type.ts:
```diff
@@ -0,0 +1,7 @@
+import { IEffect } from '../skills/effect-type'
```
Same as above about file name.

C0ZEN commented on src/types/skills/effect-type.ts:
```diff
@@ -0,0 +1,6 @@
+export interface IEffect {
```
Same as above about file name.

C0ZEN reviewed (changes requested):
Added comments.

Commit:
fix(Stat names enum): Puts AGONIZING_SNAKE_CASE.

Commit:
fix(Types): Removes the 'type' flag.

From all types.

Commit:
fix(Character command service): Adds 'service' in the name.

Commit:
fix(Character service): Changes get and set.

Removes useless 'null' type on get.
Removes useless return and variable on set.

Commit:
fix(TSconfig): Removes a wtf moment.

Commit:
fix(Character service): Puts the null back.
